crates/wascc-provider/src/states/starting.rs

```diff
@@ -0,0 +1,178 @@
+    port_assigned: i32,
+) -> anyhow::Result<ContainerHandle<ActorHandle, LogHandleFactory>> {
+    let env = <WasccProvider as Provider>::env_vars(&container, &pod, &pod_state.client).await;
+    let volume_bindings: Vec<VolumeBinding> =
+        if let Some(volume_mounts) = container.volume_mounts().as_ref() {
+            volume_mounts
+                .iter()
+                .map(|vm| -> anyhow::Result<VolumeBinding> {
+                    // Check the volume exists first
+                    let vol = pod_state.run_context.volumes.get(&vm.name).ok_or_else(|| {
+                        anyhow::anyhow!(
+                            "no volume with the name of {} found for container {}",
+                            vm.name,
+                            container.name()
+                        )
+                    })?;
+                    // We can safely assume that this should be valid UTF-8 because it would have
+                    // been validated by the k8s API
+                    Ok(VolumeBinding {
+                        name: vm.name.clone(),
+                        host_path: vol.deref().clone(),
+                    })
+                })
+                .collect::<anyhow::Result<_>>()?
+        } else {
+            vec![]
+        };
+
+    debug!("Starting container {} on thread", container.name());
+
+    let module_data = pod_state
+        .run_context
+        .modules
+        .remove(container.name())
+        .expect("FATAL ERROR: module map not properly populated");
+    let lp = pod_state.log_path.clone();
+    let host = pod_state.host.clone();
+    tokio::task::spawn_blocking(move || {
+        wascc_run_http(host, module_data, env, volume_bindings, &lp, port_assigned)
+    })
+    .await?
+}
+
+state!(
+    /// The Kubelet is starting the Pod.
+    Starting,
+    PodState,
+    Running,
+    Error,
+    {
+        info!("Starting containers for pod {:?}", pod.name());
+
+        let mut container_handles = HashMap::new();
+        for container in pod.containers() {
+            let port_assigned =
+                assign_container_port(Arc::clone(&pod_state.port_map), &pod, &container)
+                    .await
+                    .unwrap();
+            debug!(
+                "New port assigned to {} is: {}",
+                container.name(),
+                port_assigned
+            );
+
+            let container_handle = start_container(pod_state, &container, &pod, port_assigned)
+                .await
+                .unwrap();
+            container_handles.insert(
+                ContainerKey::App(container.name().to_string()),
+                container_handle,
+            );
+        }
+
+        let pod_handle = Handle::new(container_handles, pod.clone(), None).await?;
+        pod_state.handle = Some(pod_handle);
+
+        info!("All containers started for pod {:?}.", pod.name());
+
+        Ok(Transition::Advance(Running))
+    },
+    { make_status(Phase::Pending, "Starting") }
```

Comment:
Should this be `Phase::Running`? Pending is generally only used during scheduling and image pull

Reply:
I'm not sure what Kubernetes does. I figure it will mark it as running as soon as all the containers start and it transitions to running, which was what I expected. Note that the status is currently applied when it enters the state, not when it exits.

Reply:
> Note that the status is currently applied when it enters the state, not when it exits

I missed this when I went through the code